File: backend/src/routes/recipes.js
const express = require("express");
const { getDb } = require("../db");

const router = express.Router();

function parseSteps(row) {
  let steps = [];
  if (row.steps) {
    try { steps = JSON.parse(row.steps); } catch { steps = []; }
  }
  return { ...row, steps };
}

// GET /api/recipes?community=1  — public community recipes only
// GET /api/recipes               — all (built-in + user custom)
router.get("/", async (req, res, next) => {
  try {
    const db = getDb();
    const { community } = req.query;
    let rows;
    if (community === "1" || community === "true") {
      const result = await db.execute(
        "SELECT * FROM recipes WHERE isPublic = 1 ORDER BY createdAt DESC"
      );
      rows = result.rows;
    } else if (req.user) {
      // Logged in: show built-in + user's own recipes
      const result = await db.execute({
        sql: "SELECT * FROM recipes WHERE isBuiltIn = 1 OR userId = ? ORDER BY isBuiltIn DESC, createdAt DESC",
        args: [req.user.id]
      });
      rows = result.rows;
    } else {
      // Anonymous: show built-in + all non-public recipes (backward compat)
      const result = await db.execute(
        "SELECT * FROM recipes WHERE isPublic = 0 OR isBuiltIn = 1 ORDER BY isBuiltIn DESC, createdAt DESC"
      );
      rows = result.rows;
    }
    res.json(rows.map(parseSteps));
  } catch (err) { next(err); }
});

// GET /api/recipes/random — return one random recipe (optionally filtered)
router.get("/random", async (req, res, next) => {
  try {
    const db = getDb();
    const { brewerType, roastLevel, coffeeBrand } = req.query;

    const conditions = ["(isPublic = 0 OR isBuiltIn = 1)"];
    const args = [];

    if (brewerType) {
      conditions.push("brewerType = ?");
      args.push(brewerType);
    }
    if (roastLevel) {
      conditions.push("roastLevel = ?");
      args.push(roastLevel);
    }
    if (coffeeBrand) {
      conditions.push("coffeeBrand = ?");
      args.push(coffeeBrand);
    }

    const where = conditions.join(" AND ");
    const { rows } = await db.execute({
      sql: `SELECT * FROM recipes WHERE ${where} ORDER BY RANDOM() LIMIT 1`,
      args,
    });

    if (!rows[0]) return res.status(404).json({ error: "No recipes match the filters" });
    res.json(parseSteps(rows[0]));
  } catch (err) { next(err); }
});

// GET /api/recipes/liked — return array of recipe IDs the user has liked
router.get("/liked", async (req, res, next) => {
  try {
    if (!req.user) return res.json([]);
    const db = getDb();
    const { rows } = await db.execute({
      sql: "SELECT recipeId FROM recipe_likes WHERE userId = ?",
      args: [req.user.id]
    });
    res.json(rows.map(r => r.recipeId));
  } catch (err) { next(err); }
});

// POST /api/recipes/:id/like — like a recipe
router.post("/:id/like", async (req, res, next) => {
  try {
    if (!req.user) return res.status(401).json({ error: "Login required" });
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const now = new Date().toISOString();
    await db.execute({
      sql: "INSERT OR IGNORE INTO recipe_likes (userId, recipeId, createdAt) VALUES (?, ?, ?)",
      args: [req.user.id, id, now]
    });
    res.json({ liked: true });
  } catch (err) { next(err); }
});

// DELETE /api/recipes/:id/like — unlike a recipe
router.delete("/:id/like", async (req, res, next) => {
  try {
    if (!req.user) return res.status(401).json({ error: "Login required" });
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    await db.execute({
      sql: "DELETE FROM recipe_likes WHERE userId = ? AND recipeId = ?",
      args: [req.user.id, id]
    });
    res.json({ liked: false });
  } catch (err) { next(err); }
});

// GET /api/recipes/:id
router.get("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows } = await db.execute({ sql: "SELECT * FROM recipes WHERE id = ?", args: [id] });
    if (!rows[0]) return res.status(404).json({ error: "Recipe not found" });
    res.json(parseSteps(rows[0]));
  } catch (err) { next(err); }
});

// POST /api/recipes
router.post("/", async (req, res, next) => {
  try {
    const db = getDb();
    const { name, brewerType, grindSize, coffeeGrams, waterGrams, waterTempC,
            bloomTimeSec, targetBrewTimeSec, steps, sourceRecipe, notes,
            isPublic, authorName, authorSetup,
            roastLevel, coffeeBrand, coffeeName } = req.body;
    if (!name || !brewerType) return res.status(400).json({ error: "name and brewerType are required" });
    const now = new Date().toISOString();
    const userId = req.user ? req.user.id : null;
    const result = await db.execute({
      sql: `INSERT INTO recipes (name, brewerType, grindSize, coffeeGrams, waterGrams, waterTempC,
        bloomTimeSec, targetBrewTimeSec, steps, isBuiltIn, sourceRecipe, notes,
        isPublic, authorName, authorSetup, roastLevel, coffeeBrand, coffeeName,
        userId, createdAt, updatedAt)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      args: [name, brewerType, grindSize || null, coffeeGrams ?? null, waterGrams ?? null,
            waterTempC ?? null, bloomTimeSec ?? null, targetBrewTimeSec ?? null,
            steps ? JSON.stringify(steps) : null, sourceRecipe || null, notes || null,
            isPublic ? 1 : 0, authorName || null,
            authorSetup ? JSON.stringify(authorSetup) : null,
            roastLevel || null, coffeeBrand || null, coffeeName || null,
            userId, now, now]
    });
    const { rows: created } = await db.execute({ sql: "SELECT * FROM recipes WHERE id = ?", args: [Number(result.lastInsertRowid)] });
    res.status(201).json(parseSteps(created[0]));
  } catch (err) { next(err); }
});

// PUT /api/recipes/:id
router.put("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows: existingRows } = await db.execute({ sql: "SELECT * FROM recipes WHERE id = ?", args: [id] });
    const existing = existingRows[0];
    if (!existing) return res.status(404).json({ error: "Recipe not found" });
    if (existing.isBuiltIn) return res.status(403).json({ error: "Built-in recipes cannot be edited — fork it first" });
    if (req.user && existing.userId && existing.userId !== req.user.id) {
      return res.status(403).json({ error: "Not authorized to edit this recipe" });
    }
    const b = req.body;
    const now = new Date().toISOString();
    await db.execute({
      sql: `UPDATE recipes SET name=?, brewerType=?, grindSize=?, coffeeGrams=?, waterGrams=?, waterTempC=?,
        bloomTimeSec=?, targetBrewTimeSec=?, steps=?, sourceRecipe=?, notes=?,
        isPublic=?, authorName=?, roastLevel=?, coffeeBrand=?, coffeeName=?,
        updatedAt=? WHERE id=?`,
      args: [b.name ?? existing.name, b.brewerType ?? existing.brewerType,
            b.grindSize ?? existing.grindSize, b.coffeeGrams ?? existing.coffeeGrams,
            b.waterGrams ?? existing.waterGrams, b.waterTempC ?? existing.waterTempC,
            b.bloomTimeSec ?? existing.bloomTimeSec, b.targetBrewTimeSec ?? existing.targetBrewTimeSec,
            b.steps ? JSON.stringify(b.steps) : existing.steps,
            b.sourceRecipe ?? existing.sourceRecipe, b.notes ?? existing.notes,
            b.isPublic !== undefined ? (b.isPublic ? 1 : 0) : existing.isPublic,
            b.authorName ?? existing.authorName,
            b.roastLevel ?? existing.roastLevel, b.coffeeBrand ?? existing.coffeeBrand,
            b.coffeeName ?? existing.coffeeName,
            now, id]
    });
    const { rows: updated } = await db.execute({ sql: "SELECT * FROM recipes WHERE id = ?", args: [id] });
    res.json(parseSteps(updated[0]));
  } catch (err) { next(err); }
});

// DELETE /api/recipes/:id
router.delete("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows: existingRows } = await db.execute({ sql: "SELECT id, isBuiltIn, userId FROM recipes WHERE id = ?", args: [id] });
    const existing = existingRows[0];
    if (!existing) return res.status(404).json({ error: "Recipe not found" });
    if (existing.isBuiltIn) return res.status(403).json({ error: "Built-in recipes cannot be deleted" });
    if (req.user && existing.userId && existing.userId !== req.user.id) {
      return res.status(403).json({ error: "Not authorized to delete this recipe" });
    }
    await db.execute({ sql: "DELETE FROM recipes WHERE id = ?", args: [id] });
    res.status(204).send();
  } catch (err) { next(err); }
});

module.exports = router;

File: backend-py/app/routers/recipes.py
"""Recipes router — CRUD, community, likes, random."""

import json
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel

from ..database import get_db
from ..auth.dependencies import AuthUser, optional_auth, require_auth

router = APIRouter(prefix="/api/recipes", tags=["recipes"])


def _parse_recipe(row, columns):
    """Convert row to dict, parsing JSON steps and authorSetup."""
    d = dict(zip(columns, row))
    if d.get("steps"):
        try:
            d["steps"] = json.loads(d["steps"])
        except (json.JSONDecodeError, TypeError):
            d["steps"] = []
    else:
        d["steps"] = []
    if d.get("authorSetup"):
        try:
            d["authorSetup"] = json.loads(d["authorSetup"])
        except (json.JSONDecodeError, TypeError):
            d["authorSetup"] = None
    return d


def _get_recipe_columns(db):
    cursor = db.execute("SELECT * FROM recipes LIMIT 0")
    return [desc[0] for desc in cursor.description]


@router.get("/")
async def list_recipes(
    community: Optional[str] = None,
    user: AuthUser | None = Depends(optional_auth),
):
    db = get_db()
    columns = _get_recipe_columns(db)

    if community in ("1", "true"):
        rows = db.execute("SELECT * FROM recipes WHERE isPublic = 1 ORDER BY createdAt DESC").fetchall()
    elif user:
        rows = db.execute(
            "SELECT * FROM recipes WHERE isBuiltIn = 1 OR userId = ? ORDER BY isBuiltIn DESC, createdAt DESC",
            [user.id],
        ).fetchall()
    else:
        rows = db.execute(
            "SELECT * FROM recipes WHERE isPublic = 0 OR isBuiltIn = 1 ORDER BY isBuiltIn DESC, createdAt DESC"
        ).fetchall()

    return [_parse_recipe(r, columns) for r in rows]


@router.get("/random")
async def random_recipe(
    brewerType: Optional[str] = None,
    roastLevel: Optional[str] = None,
    coffeeBrand: Optional[str] = None,
):
    db = get_db()
    conditions = ["(isPublic = 0 OR isBuiltIn = 1)"]
    args = []

    if brewerType:
        conditions.append("brewerType = ?")
        args.append(brewerType)
    if roastLevel:
        conditions.append("roastLevel = ?")
        args.append(roastLevel)
    if coffeeBrand:
        conditions.append("coffeeBrand = ?")
        args.append(coffeeBrand)

    where = " AND ".join(conditions)
    columns = _get_recipe_columns(db)
    row = db.execute(f"SELECT * FROM recipes WHERE {where} ORDER BY RANDOM() LIMIT 1", args).fetchone()

    if not row:
        raise HTTPException(status_code=404, detail="No recipes match the filters")

    return _parse_recipe(row, columns)


@router.get("/liked")
async def liked_recipes(user: AuthUser | None = Depends(optional_auth)):
    if not user:
        return []
    db = get_db()
    rows = db.execute("SELECT recipeId FROM recipe_likes WHERE userId = ?", [user.id]).fetchall()
    return [r[0] for r in rows]


@router.post("/{recipe_id}/like")
async def like_recipe(recipe_id: int, user: AuthUser = Depends(require_auth)):
    db = get_db()
    now = datetime.now(timezone.utc).isoformat()
    db.execute(
        "INSERT OR IGNORE INTO recipe_likes (userId, recipeId, createdAt) VALUES (?, ?, ?)",
        [user.id, recipe_id, now],
    )
    db.commit()
    return {"liked": True}


@router.delete("/{recipe_id}/like")
async def unlike_recipe(recipe_id: int, user: AuthUser = Depends(require_auth)):
    db = get_db()
    db.execute("DELETE FROM recipe_likes WHERE userId = ? AND recipeId = ?", [user.id, recipe_id])
    db.commit()
    return {"liked": False}


@router.get("/{recipe_id}")
async def get_recipe(recipe_id: int):
    db = get_db()
    columns = _get_recipe_columns(db)
    row = db.execute("SELECT * FROM recipes WHERE id = ?", [recipe_id]).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Recipe not found")
    return _parse_recipe(row, columns)


class RecipeCreate(BaseModel):
    name: str
    brewerType: str
    grindSize: Optional[str] = None
    coffeeGrams: Optional[float] = None
    waterGrams: Optional[float] = None
    waterTempC: Optional[int] = None
    bloomTimeSec: Optional[int] = None
    targetBrewTimeSec: Optional[int] = None
    steps: Optional[list] = None
    sourceRecipe: Optional[str] = None
    notes: Optional[str] = None
    isPublic: Optional[bool] = False
    authorName: Optional[str] = None
    authorSetup: Optional[dict] = None
    roastLevel: Optional[str] = None
    coffeeBrand: Optional[str] = None
    coffeeName: Optional[str] = None


@router.post("/", status_code=201)
async def create_recipe(body: RecipeCreate, user: AuthUser | None = Depends(optional_auth)):
    if not body.name or not body.brewerType:
        raise HTTPException(status_code=400, detail="name and brewerType are required")

    db = get_db()
    now = datetime.now(timezone.utc).isoformat()
    user_id = user.id if user else None

    cursor = db.execute(
        """INSERT INTO recipes (name, brewerType, grindSize, coffeeGrams, waterGrams, waterTempC,
            bloomTimeSec, targetBrewTimeSec, steps, isBuiltIn, sourceRecipe, notes,
            isPublic, authorName, authorSetup, roastLevel, coffeeBrand, coffeeName,
            userId, createdAt, updatedAt)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        [
            body.name, body.brewerType, body.grindSize, body.coffeeGrams, body.waterGrams,
            body.waterTempC, body.bloomTimeSec, body.targetBrewTimeSec,
            json.dumps(body.steps) if body.steps else None,
            body.sourceRecipe, body.notes,
            1 if body.isPublic else 0, body.authorName,
            json.dumps(body.authorSetup) if body.authorSetup else None,
            body.roastLevel, body.coffeeBrand, body.coffeeName,
            user_id, now, now,
        ],
    )
    db.commit()
    recipe_id = cursor.lastrowid

    columns = _get_recipe_columns(db)
    row = db.execute("SELECT * FROM recipes WHERE id = ?", [recipe_id]).fetchone()
    return _parse_recipe(row, columns)


@router.put("/{recipe_id}")
async def update_recipe(recipe_id: int, body: dict, user: AuthUser | None = Depends(optional_auth)):
    db = get_db()
    columns = _get_recipe_columns(db)
    row = db.execute("SELECT * FROM recipes WHERE id = ?", [recipe_id]).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Recipe not found")

    existing = dict(zip(columns, row))
    if existing.get("isBuiltIn"):
        raise HTTPException(status_code=403, detail="Built-in recipes cannot be edited — fork it first")
    if user and existing.get("userId") and existing["userId"] != user.id:
        raise HTTPException(status_code=403, detail="Not authorized to edit this recipe")

    now = datetime.now(timezone.utc).isoformat()
    steps_val = json.dumps(body["steps"]) if "steps" in body and body["steps"] is not None else existing.get("steps")

    db.execute(
        """UPDATE recipes SET name=?, brewerType=?, grindSize=?, coffeeGrams=?, waterGrams=?, waterTempC=?,
            bloomTimeSec=?, targetBrewTimeSec=?, steps=?, sourceRecipe=?, notes=?,
            isPublic=?, authorName=?, roastLevel=?, coffeeBrand=?, coffeeName=?,
            updatedAt=? WHERE id=?""",
        [
            body.get("name", existing["name"]),
            body.get("brewerType", existing["brewerType"]),
            body.get("grindSize", existing.get("grindSize")),
            body.get("coffeeGrams", existing.get("coffeeGrams")),
            body.get("waterGrams", existing.get("waterGrams")),
            body.get("waterTempC", existing.get("waterTempC")),
            body.get("bloomTimeSec", existing.get("bloomTimeSec")),
            body.get("targetBrewTimeSec", existing.get("targetBrewTimeSec")),
            steps_val,
            body.get("sourceRecipe", existing.get("sourceRecipe")),
            body.get("notes", existing.get("notes")),
            (1 if body["isPublic"] else 0) if "isPublic" in body else existing.get("isPublic"),
            body.get("authorName", existing.get("authorName")),
            body.get("roastLevel", existing.get("roastLevel")),
            body.get("coffeeBrand", existing.get("coffeeBrand")),
            body.get("coffeeName", existing.get("coffeeName")),
            now, recipe_id,
        ],
    )
    db.commit()

    updated_row = db.execute("SELECT * FROM recipes WHERE id = ?", [recipe_id]).fetchone()
    return _parse_recipe(updated_row, columns)


@router.delete("/{recipe_id}", status_code=204)
async def delete_recipe(recipe_id: int, user: AuthUser | None = Depends(optional_auth)):
    db = get_db()
    row = db.execute("SELECT id, isBuiltIn, userId FROM recipes WHERE id = ?", [recipe_id]).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Recipe not found")
    if row[1]:  # isBuiltIn
        raise HTTPException(status_code=403, detail="Built-in recipes cannot be deleted")
    if user and row[2] and row[2] != user.id:
        raise HTTPException(status_code=403, detail="Not authorized to delete this recipe")

    db.execute("DELETE FROM recipes WHERE id = ?", [recipe_id])
    db.commit()
    return Response(status_code=204)
